src/utils/format.ts

```diff
@@ -0,0 +1,269 @@
+const SqlString = require('../sql-string');
+
+const operatorsSet = new Set(Object.values(operators));
+
+export function format(arr: unknown[], dialect: string): string {
+  const timeZone = null;
+
+  // Make a clone of the array because format modifies the passed args
+  return SqlString.format(arr[0], arr.slice(1), timeZone, dialect);
+}
+
+export function formatNamedParameters(
+  sql: string,
+  parameters: Record<string, unknown>,
+  dialect: string,
+): string {
+  return SqlString.formatNamedParameters(sql, parameters, null, dialect);
+}
+
+export type FinderOptions<TAttributes> = {
+  attributes?: string[],
+  where?: WhereOptions<TAttributes>,
+};
+
+export type MappedFinderOptions<TAttributes> = Omit<FinderOptions<TAttributes>, 'attributes'> & {
+  // an array of attribute-column mapping, or just attributes
+  attributes?: Array<[columnName: string, attributeName: string] | string>,
+};
+
+/**
+ * Expand and normalize finder options.
+ * Mutates the "options" parameter.
+ *
+ * @param options
+ * @param Model
+ */
+export function mapFinderOptions<M extends Model, T extends FinderOptions<Attributes<M>>>(
+  options: T,
+  Model: ModelStatic<Model>,
+): MappedFinderOptions<Attributes<M>> {
+  if (Array.isArray(options.attributes)) {
+    options.attributes = Model._injectDependentVirtualAttributes(
+      options.attributes,
+    );
+
+    options.attributes = options.attributes.filter(
+      v => !Model._virtualAttributes.has(v),
+    );
+  }
+
+  mapOptionFieldNames(options, Model);
+
+  return options;
+}
+
+/**
+ * Used to map field names in attributes and where conditions.
+ *
+ * Mutates the "options" parameter.
+ *
+ * @param options
+ * @param Model
+ */
+export function mapOptionFieldNames<M extends Model>(
+  options: FinderOptions<Attributes<M>>,
+  Model: ModelStatic<Model>,
+): MappedFinderOptions<Attributes<M>> {
+
+  // note: parts of Sequelize rely on this function mutating its inputs.
+  //  be aware that these places need to be fixed before trying to make this a pure function.
+  //  - ephys
+
+  const out: MappedFinderOptions<Attributes<M>> = options;
+
+  if (Array.isArray(options.attributes)) {
+    out.attributes = options.attributes.map(attributeName => {
+      // Object lookups will force any variable to strings, we don't want that for special objects etc
+      if (typeof attributeName !== 'string') {
+        return attributeName;
+      }
+
+      // Map attributes to column names
+      const columnName: string | undefined = Model.rawAttributes[attributeName]?.field;
+      if (columnName && columnName !== attributeName) {
+        return [columnName, attributeName];
+      }
+
+      return attributeName;
+    });
+  }
+
+  if (options.where != null && isPlainObject(options.where)) {
+    // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error -- this fails in TS 4.4 and up, but not before
+    // @ts-ignore the work necessary to type the return type of mapWhereFieldNames is not worth it
```

Comment:
I think it's something to evaluate as we migrate parts of the codebase that use `mapWhereFieldNames`. If it's useful to these parts, let's migrate. Otherwise the complex typings it would require aren't worth it